File: apps/ios/ExpCore/Sources/API/RelationsApi.swift
import Foundation

// Mirrors apps/web/src/lib/trpc/relations.ts (EXP-736) — typed links between
// two issues. The server owns the canonical direction: `inverse` says the
// CALLER's issue is the related side, so it swaps the pair before writing.
public struct CreateRelationInput: Encodable, Sendable {
    public let issueId: String
    public let relatedIssueId: String
    /// An `IssueRelationType` raw value.
    public let type: String
    public let inverse: Bool

    public init(issueId: String, relatedIssueId: String, type: String, inverse: Bool) {
        self.issueId = issueId
        self.relatedIssueId = relatedIssueId
        self.type = type
        self.inverse = inverse
    }
}

public struct DeleteRelationInput: Encodable, Sendable {
    /// The `issue_relations` row id.
    public let id: String

    public init(id: String) {
        self.id = id
    }
}

public final class RelationsApi: Sendable {
    private let trpc: TrpcClient

    public init(trpc: TrpcClient) {
        self.trpc = trpc
    }

    /// Both mutations return `{txId}` only — the row itself arrives over the
    /// `issue-relations` shape, so the body is ignored.
    public func create(accountId: String, _ input: CreateRelationInput) async throws {
        try await trpc.mutationVoid(accountId: accountId, path: "relations.create", input: input)
    }

    public func delete(accountId: String, id: String) async throws {
        try await trpc.mutationVoid(accountId: accountId, path: "relations.delete", input: DeleteRelationInput(id: id))
    }
}

File: apps/android/app/src/main/java/com/exponential/app/data/api/RelationsApi.kt
package com.exponential.app.data.api

import javax.inject.Inject
import javax.inject.Singleton
import kotlinx.serialization.Serializable

/**
 * Input for `relations.create` (EXP-736). The row is always stored in the
 * CANONICAL direction — `inverse` flips the two ids server-side, so a
 * "Blocked by" pick sends the same `blocks` type as "Blocking" does.
 */
@Serializable
data class CreateRelationInput(
    val issueId: String,
    val relatedIssueId: String,
    val type: String,
    val inverse: Boolean = false,
)

@Serializable
data class DeleteRelationInput(val id: String)

// Mirrors apps/web/src/lib/trpc/relations.ts. Marking a DUPLICATE never comes
// through here: it stays the issues.update({duplicateOfId}) lockstep write
// (status + FK + mirror row in one transaction), which is why the picker's
// "Duplicate of" entry routes to IssuesApi.setDuplicateOf instead.
@Singleton
class RelationsApi @Inject constructor(private val trpc: TrpcClient) {

    suspend fun create(accountId: String, input: CreateRelationInput) {
        trpc.mutationUnit(
            accountId,
            path = "relations.create",
            input = input,
            inputSerializer = CreateRelationInput.serializer(),
        )
    }

    suspend fun delete(accountId: String, id: String) {
        trpc.mutationUnit(
            accountId,
            path = "relations.delete",
            input = DeleteRelationInput(id),
            inputSerializer = DeleteRelationInput.serializer(),
        )
    }
}
